Common/Orders/Fills/EquityFillModel.cs

```diff
@@ -461,127 +503,310 @@ public virtual OrderEvent MarketOnCloseFill(Security asset, MarketOnCloseOrder o
+                    endTime = quote.EndTime;
+                    return quote.BidPrice;
                 }
             }
-            return true;
+
+            if (subscribedTypes.Contains(typeof(QuoteBar)))
+            {
+                var quoteBar = asset.Cache.GetData<QuoteBar>();
+                if (quoteBar != null)
+                {
+                    endTime = quoteBar.EndTime;
+                    return quoteBar.Bid?.Close ?? quoteBar.Close;
+                }
+            }
+
+            if (isTickSubscribed)
+            {
+                var trade = ticks.LastOrDefault(x => x.TickType == TickType.Trade && x.Price != 0);
+                if (trade != null)
+                {
+                    endTime = trade.EndTime;
+                    return trade.Price;
+                }
+            }
+
+            if (subscribedTypes.Contains(typeof(TradeBar)))
+            {
+                var tradeBar = asset.Cache.GetData<TradeBar>();
+                if (tradeBar != null)
+                {
+                    endTime = tradeBar.EndTime;
+                    return tradeBar.Close;
+                }
+            }
+
+            endTime = asset.LocalTime;
+            return asset.BidPrice;
         }
 
-        public class Prices
+        /// <summary>
+        /// Get current trade bar for subscribed data
+        /// </summary>
+        /// <param name="asset">Security which has subscribed data types</param>
+        private TradeBar GetLastTradeBar(Security asset)
         {
-            public readonly DateTime EndTime;
-            public readonly decimal Current;
-            public readonly decimal Open;
-            public readonly decimal High;
-            public readonly decimal Low;
-            public readonly decimal Close;
-
-            public Prices(IBaseDataBar bar)
-                : this(bar.EndTime, bar.Close, bar.Open, bar.High, bar.Low, bar.Close)
+            var subscribedTypes = GetSubscribedTypes(asset);
+
+            if (subscribedTypes.Contains(typeof(Tick)))
             {
+                var ticks = asset.Cache.GetAll<Tick>().ToList();
+
+                var trade = ticks.LastOrDefault(x => x.TickType == TickType.Trade && x.Price != 0);
+                if (trade != null)
+                {
+                    return new TradeBar(trade.Time, trade.Symbol, trade.Price, trade.Price, trade.Price, trade.Price, trade.Quantity, TimeSpan.Zero);
+                }
+
+                DateTime quoteTime;
+                decimal askPrice;
+                decimal askSize;
+                decimal bidPrice;
+                decimal bidSize;
+
+                if (TryGetLastQuoteTick(asset, ticks, out quoteTime, out bidPrice, out bidSize, out askPrice, out askSize))
+                {
+                    var price = (askPrice + bidPrice) / 2;
+                    return new TradeBar(quoteTime, asset.Symbol, price, price, price, price, 0, TimeSpan.Zero);
+                }
```

Comment:
This will falsely report a trade for fill methods that rely on trades to trigger.